### PR TITLE
refactor(compiler-cli): report better error when a `ModuleWithProviders` call occurs within `Component.imports`

### LINKED ISSUE #46003: Error for importing ModuleWithProviders into @Component.imports could be improved
### Which @angular/* package(s) are the source of the bug?

compiler

### Is this a regression?

No

### Description

Using `RouterModule.forRoot([...])` in standalone's component imports results in errors that are not immediately actionable. This is a pretty common scenario since people trying out "standalone" will try to move router's configuration _somewhere_ and many will try moving the `RouterModule.forRoot([...])` into `@Component.imports`., ex.:

```typescript
@Component({
  selector: 'app-root',
  template: `
   ....
  `,
  standalone: true,
  imports: [RouterModule.forRoot([])],
})
export class AppRoot {
}
```

Trying to do the above will result in the 2 following errors:

```
Error: src/main.ts:25:12 - error NG1010: 'imports' must be an array of components, directives, pipes, or NgModules
  Value is of type '[(not statically analyzable)]'.

25   imports: [RouterModule.forRoot([])],
              ~~~~~~~~~~~~~~~~~~~~~~~~~~


Error: src/main.ts:25:13 - error TS2322: Type 'ModuleWithProviders<RouterModule>' is not assignable to type 'any[] | Type<any>'.

25   imports: [RouterModule.forRoot([])],
               ~~~~~~~~~~~~~~~~~~~~~~~~
``` 

Things that might be confusing to the users:
* `(not statically analyzable)` might be cryptic and not-actionable
* the second error actually gets us a bit more info since it indicates that `ModuleWithProviders` don't "fit" into the `@Component.imports` "slot"

Not sure if we can improve things on the compiler side, especially for the first error. 

### Please provide the environment you discovered this bug in (run `ng version`)

```true
v14
```

### PR DESCRIPTION
**Pull request by JoostK**
See individual commits

Closes #46003

### CODE REVIEW

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM but would be great to have another pass from someone more familiar with the compiler (@alxhub ?)

File: packages/compiler-cli/test/ngtsc/standalone_spec.ts [hunk 1]
```diff
@@ -274,6 +274,55 @@ runInEachFileSystem(() => {
...
+            template: '<div></div>',
+            standalone: true,
+            imports: [TestModule.forRoot()],
+          })
+          export class TestCmp {}
+        `);
+        const diags = env.driveDiagnostics();
+        expect(diags.length).toBe(1);
+        expect(diags[0].code).toBe(ngErrorCode(ErrorCode.COMPONENT_UNKNOWN_IMPORT));
+        expect(getSourceCodeForDiagnostic(diags[0])).toEqual('[TestModule.forRoot()]');
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/test/ngtsc/standalone_spec.ts:418]: Good point, I agree; I've added message assertions for these tests.

File: packages/compiler-cli/src/ngtsc/annotations/component/src/util.ts [hunk 2]
```diff
@@ -92,6 +94,17 @@ export function validateAndFlattenComponentImports(imports: ResolvedValue, expr:
...
+      let origin = expr;
+      if (ref instanceof SyntheticValue) {
+        // The `ModuleWithProviders` type originated from a foreign function declaration, in which
+        // case the original foreign call is available which is used to get a more accurate origin
+        // node that points at the specific call expression.
+        origin = getOriginNodeForDiagnostics(ref.value.mwpCall, expr);
+      }
+      diagnostics.push(makeDiagnostic(
+          ErrorCode.COMPONENT_UNKNOWN_IMPORT, origin,
+          `'imports' contains a module with providers. Modules with providers are not supported in standalone component imports`));
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/component/src/util.ts:107]: How about wording like:
```suggestion
          `'imports' contains a ModuleWithProviders value, likely the result of a 'Module.forRoot()'-style call. These calls are not used to configure components and are not valid in standalone component imports - consider importing them in the application bootstrap instead.`));
```

File: packages/compiler-cli/test/ngtsc/standalone_spec.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/standalone_spec.ts:418]: I actually agree with the first part of Pawel's statement - including specific error messages in the test assertion adds a lot of brittleness and we should not do that.

A good compromise and a good best practice in general would be to include an example of the error experience in the commit message itself. An easy way to generate this is to temporarily replace `env.driveDiagnostics()` with `env.driveMain()` and copy-paste the resulting diagnostic in its context.

**Comment by JoostK** [comment]
caretaker note: please approve one more time

**Review by alxhub — APPROVED** [review]

**Comment by alxhub** [comment]
This PR was merged into the repository by commit e592cc7c1cc86aa32750bb1054c0e7a2899ac1f8.

**Pull request closed without merging**